rfc/20240610-Tofu-Version-Compatibility.md

```diff
@@ -0,0 +1,106 @@
+
+
+What happens if only the `version.tf` file exists in this module with no tofu overwrite? This is a common case for modules who have not yet considered explicit OpenTofu migration.
+
+As a user, I would like to know if any incompatible terraform_version has been specified as part of my project's configuration and modules. I would also assume any version requirement under 1.6 would be identical between Terraform and OpenTofu.
+
+```hcl
+terraform {
+    required_version = 1.7.4
+}
+```
+```shell
+$ TF_LOG=debug tofu init
+Initializing the backend...
+Initializing modules...
+Warning: Configuration requests terraform version >= 1.6!
+  Please update the configuration to specify the required OpenTofu Version.  More information is available in the debug log.
+Debug: terraform required_version = 1.7.4 in main.tf:421
+```
+
+This could also include some information on how to remedy this situation, perhaps linking to the docs.
+
+### Technical Approach
+
+The "required_version" field is accessed via `sniffCoreVersionRequirements(body)` in internal/configs/parser_config.go and stored in `configs.File.CoreVersionConstraints`. These are then merged into `configs.Module.CoreVersionConstraints` and are checked in `configs.Module.CheckCoreVersionRequirements()`. This check function is called in a variety of locations moving up the stack, but is as high as we need to concern ourselves with for this RFC. We will need to make sure the warning won't be emitted multiple times due to the spaghettified command package.
+
+We have two very similar paths here:
+* Split CoreVersionConstraints into two fields: `TofuVersionConstraints` and `TerraformVersionConstraints`
+* Add a flag into each VersionConstraint that understands if it is Tofu or Terraform specific.
+
+Either path is easy to both implement and thoroughly test.
+
+The `configs.Module.CheckCoreVersionRequirements()` function is where we will need to emit the warnings and corresponding debug log entries.
+
+### Open Questions
+
+Should there be an option (config, CLI, or ENV) to:
+* Disable the version checking altogether?
+* Don't check terraform required_versions
+* Treat both Tofu and Terraform version as identical instead of using our equivalent "terraform version" guess?
+* Do we show the guessed terraform version in `tofu -version`, maybe with a `-verbose` flag?
+
+### Future Considerations
+
+If we run into providers expecting a particular terraform version (even though this is deprecated), we may want to use the "Terraform Version Equivalent" in that API call.
+
+## Potential Alternatives
+
+Don't implement this and assume that all module and project authors will adopt the .tofu extension and keep the respective required_versions up to date.
+
+Implement an option to disable required_version checking.
+
+Try to guess the equivalent OpenTofu version for a specified Terraform version.
+
```

Comment:
Building a GitHub action was proposed as a possible alternative action as well. I particularly like this no matter what solution we land on because I'm assuming I can use it to guard against unexpected incompatibilities in modules I include in my code as well when I update versions.

Reply:
We also discussed some sort of compatibility table between versions and modules based on real user feedback. I know it's tricky, but still, it has some practical benefits. I would add those to the open questions block

Reply:
Noted in b3b094db80